libs/api/ebpf_api.cpp

```diff
@@ -4920,18 +4929,123 @@ CATCH_NO_MEMORY_EBPF_RESULT
 _Must_inspect_result_ ebpf_result_t
 ebpf_perf_event_array_map_subscribe(
     fd_t map_fd,
+    uint32_t cpu_id,
     _Inout_opt_ void* callback_context,
     perf_buffer_sample_fn sample_callback,
     perf_buffer_lost_fn lost_callback,
     _Outptr_ perf_event_array_subscription_t** subscription) NO_EXCEPT_TRY
 {
     EBPF_LOG_ENTRY();
-    UNREFERENCED_PARAMETER(map_fd);
-    UNREFERENCED_PARAMETER(callback_context);
-    UNREFERENCED_PARAMETER(sample_callback);
-    UNREFERENCED_PARAMETER(lost_callback);
-    UNREFERENCED_PARAMETER(subscription);
-    EBPF_RETURN_RESULT(EBPF_OPERATION_NOT_SUPPORTED);
+
+    try {
+
+        ebpf_assert(sample_callback);
+        ebpf_assert(subscription);
+        ebpf_result_t result = EBPF_SUCCESS;
+
+        uint32_t key_size = 0;
+        uint32_t value_size = 0;
+        uint32_t max_entries = 0;
+        uint32_t type;
+
+        ebpf_handle_t map_handle = _get_handle_from_file_descriptor(map_fd);
+        if (map_handle == ebpf_handle_invalid) {
+            result = EBPF_INVALID_FD;
+            EBPF_RETURN_RESULT(result);
+        }
+
+        result = _get_map_descriptor_properties(map_handle, &type, &key_size, &value_size, &max_entries);
+        if (result != EBPF_SUCCESS) {
+            EBPF_RETURN_RESULT(result);
+        }
+
+        if (type != BPF_MAP_TYPE_PERF_EVENT_ARRAY) {
+            result = EBPF_INVALID_ARGUMENT;
+            EBPF_LOG_MESSAGE_ERROR(
+                EBPF_TRACELOG_LEVEL_ERROR,
+                EBPF_TRACELOG_KEYWORD_API,
+                "perf_buffer__new API is called on a map that is not of the perf event Array type.",
+                result);
+            EBPF_RETURN_RESULT(result);
+        }
+
+        *subscription = nullptr;
+
+        ebpf_perf_event_array_subscription_ptr local_subscription =
+            std::make_unique<ebpf_perf_event_array_subscription_t>();
+        local_subscription->perf_event_array_map_handle = ebpf_handle_invalid;
+
+        if (!Platform::DuplicateHandle(
+                reinterpret_cast<ebpf_handle_t>(GetCurrentProcess()),
+                map_handle,
+                reinterpret_cast<ebpf_handle_t>(GetCurrentProcess()),
+                &local_subscription->perf_event_array_map_handle,
```

Comment:
do we need to manually clean up the local_subscription->perf_event_array_map_handle in the case of failure after this duplication succeeds? (I.e any of the failure cases in the below processing)

Reply:
The destructor takes care of freeing it.  It is a smart pointer.   
~_ebpf_perf_event_array_subscription()
 {
     EBPF_LOG_ENTRY();

     if (async_ioctl_completion != nullptr) {
         clean_up_async_ioctl_completion(async_ioctl_completion);
     }
     if (perf_event_array_map_handle != ebpf_handle_invalid) {
         Platform::CloseHandle(perf_event_array_map_handle);
     }

     EBPF_LOG_EXIT();
 }